### PR TITLE
crimson: refactor osd operation pipeline

### PR DESCRIPTION
**Pull request by athanatos**
The primary goal here is to pull all pg routed messages into a common start_pg_operation pathway that will later be generalized to deal with routing to pgs on different reactors.

Also simplifies the op ordering machinery.

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by athanatos** [comment]
@rzarzynski I've also got a version of this rebased on your other branch, but it looks like that one still has a little way to go?

**Comment by rzarzynski** [comment]
@athanatos: it requires polishing (human-readable names for events, visits time etc.) but I think the core stuff is quite ready. I pushed today the exit barrier tracking and restructured the commit order to have WIPs / DNMs on top of the branch.

**Comment by athanatos** [comment]
@rzarzynski I think this PR is ready to go, but I can rebase it on yours if you'd prefer that that go in first.

**Comment by rzarzynski** [comment]
@athanatos: the changes here are pretty broad. I guess it might be less work in total if we leverage the existing refactor-on-top-of-tracking than the opposite. If that's the case, I've repushed op tracking without WIPs nor DNMs.

Anyway, I really don't have strong feelings here and would be perfectly fine with merging your branch first.

**Comment by athanatos** [comment]
Ok, I'll take a look at your branch tomorrow.

**Comment by athanatos** [comment]
@rzarzynski Rebased.  https://github.com/ceph/ceph/pull/46194 has some bugs I noticed during testing.

**Comment by rzarzynski** [comment]
```
99% tests passed, 1 tests failed out of 277

Total Test time (real) = 2169.17 sec

The following tests FAILED:
	 31 - run-rbd-unit-tests-1.sh (Failed)
```

**Comment by rzarzynski** [comment]
jenkins retest this please

**Comment by rzarzynski** [comment]
@athanatos: reviewed. Thanks for catching them up!

**Comment by athanatos** [comment]
@rzarzynski Did you mean that you reviewed this PR?  If so, can you approve it?

File: src/crimson/osd/osd_operations/client_request.cc [hunk 1]
```diff
@@ -23,12 +20,42 @@ namespace {
 
 namespace crimson::osd {
 
+
+void ClientRequest::Orderer::requeue(
+  ShardServices &shard_services, Ref<PG> pg)
+{
+  for (auto &req: list) {
+    req.handle.exit();
```
Comment (rzarzynski) [inline src/crimson/osd/osd_operations/client_request.cc:28]: Way simpler!

**Review by rzarzynski — APPROVED** [review]

**Pull request merged**